Feature/2214 extract adapt api quickstart as ngsi ld tuturial (#6)
stellio-hub/stellio-docs: docs/API_walkthrough.md

# Quickstart

This quickstart guide shows a real use case scenario of interaction with the API in an Apiculture context.

## Prepare your environment

The provided examples make use of the HTTPie command line tool (installation instructions: https://httpie.org/docs#installation)

> All requests are grouped in a Postman collection that can be found [here](samples/API_Quick_Start.postman_collection.json).

Reviewer: this file will have to be moved

Reviewer: Could be nice, no?

Author: Yes, added

For more details about how to import a Postman collection see https://learning.postman.com/docs/getting-started/importing-and-exporting-data/.

> Start a Stellio instance. You can use the provided Docker compose configuration in this directory:

Reviewer: you'll have to reference the Stellio repository

```json
{
   "id": "urn:ngsi-ld:BeeHive:01",
   "type": "BeeHive",
   "belongs": {
      "type": "Relationship",
      "object": "urn:ngsi-ld:Apiary:01"
   },
   "managedBy": {
     "type": "Relationship",
     "object": "urn:ngsi-ld:Beekeeper:01"
   },
   "location": {
      "type": "GeoProperty",
      "value": {
         "type": "Point",
         "coordinates": [
            24.30623,
            60.07966
         ]
      }
   },
   "temperature": {
      "type": "Property",
      "value": 22.2,
      "unitCode": "CEL",
      "observedAt": "2019-10-26T21:32:52.98601Z",
      "observedBy": {
         "type": "Relationship",
         "object": "urn:ngsi-ld:Sensor:01"
      }
   },
   "humidity": {
      "type": "Property",
      "value": 60,
      "unitCode": "P1",
      "observedAt": "2019-10-26T21:32:52.98601Z",
      "observedBy": {
         "type": "Relationship",
         "object": "urn:ngsi-ld:Sensor:02"
      }
   },
   "@context": [
      "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld"
   ]
}
```

## Queries

* We start by creating the beekeeper, the apiary, the beehive and sensors:

```shell

> http POST http://localhost:8080/ngsi-ld/v1/entities Content-Type:application/ld+json < Beekeeper.jsonld

Reviewer: it would be nice to present them in such a way: https://httpie.io/docs#raw-and-complex-json (users could then just copy / paste the commands without having to create intermediate files with sample data)

    "typeList": [
        "https://ontology.eglobalmark.com/apic#Beekeeper",
        "https://ontology.eglobalmark.com/apic#BeeHive",
        "https://ontology.eglobalmark.com/apic#Apiary",
        "https://ontology.eglobalmark.com/egm#Sensor"
    ]
}
```

* To get more details about a type of entity, a Beekeeper for example:

```shell
http http://localhost:8080/ngsi-ld/v1/entities type==Beekeeper Link:$CONTEXT_LINK
```

Sample payload returned showing more details about the entity Beekeeper:

```json
{
    "id": "https://ontology.eglobalmark.com/apic#Beekeeper",
    "type": "EntityTypeInfo",
    "typeName": "Beekeeper",
    "entityCount": 1,
    "attributeDetails": [
        {
            "id": "https://uri.etsi.org/ngsi-ld/name",
            "type": "Attribute",
            "attributeName": "name",
            "attributeTypes": [
                "Property"
            ]
        }
    ]
}
```

* The created beehive can be retrieved by id:

```shell
http http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01 Link:$CONTEXT_LINK
```

* Or by querying all entities of type BeeHive:

```shell
http http://localhost:8080/ngsi-ld/v1/entities type==BeeHive Link:$CONTEXT_LINK
```

> * Adding the options `KeyValues` to the parameter of a get query will return a reduced version of the Entity providing only top level attribute and their value or object.

Reviewer: ```suggestion
* Adding the options `keyValues` to the request parameters will return a reduced version of the entity providing only top level attribute and their value or object.
```

```shell
http POST http://localhost:8080/ngsi-ld/v1/subscriptions Content-Type:application/ld+json < subscription_to_beehive.jsonld
```

subscription_to_beehive.jsonld
```json
{
  "id":"urn:ngsi-ld:Subscription:01",
  "type":"Subscription",
  "entities": [
    {
      "type": "BeeHive"
    }
  ],
  "q": "temperature>40",
  "notification": {
    "attributes": ["temperature"],
    "format": "normalized",
    "endpoint": {
      "uri": "http://ptsv2.com/t/ir6uu-1626449561/post",
      "accept": "application/json"
    }
  },
  "@context": [
     "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld"
  ]
}
```

* The created subscription can be retrieved by id:

```shell
http http://localhost:8080/ngsi-ld/v1/subscriptions/urn:ngsi-ld:Subscription:01 Link:$CONTEXT_LINK
```

* We can update the temperature with a value grater then 40, using the following query:

```shell

> http PATCH http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs \

Reviewer: maybe a partial attribute update?

Reviewer: no answer?

Author: Didn't got it (I thought that it was about the title of the section)

                "createdAt": "2021-07-16T15:51:54.215300Z",
                "value": 42,
                "observedAt": "2019-10-26T22:35:52.986010Z",
                "unitCode": "CEL"
            },
            "@context": [
                "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld"
            ]
        }
    ]
}
```

* We can also update the beehive humidity:

```shell
http PATCH http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs \
    Link:$CONTEXT_LINK < beehive_updateHumidity.json
```

beehive_updateHumidity.json
```json
{
   "humidity": {
     "type": "Property",
     "value": 58,
     "unitCode": "P1",
     "observedAt": "2019-10-26T21:35:52.98601Z",
     "observedBy": {
        "type": "Relationship",
        "object": "urn:ngsi-ld:Sensor:02"
     }
  }
}
```

* Since we updated both temperature and humidity, we can get the temporal evolution of those properties

```shell
http http://localhost:8080/ngsi-ld/v1/temporal/entities/urn:ngsi-ld:BeeHive:01 \
    timerel==between \
    time==2019-10-25T12:00:00Z \
    endTime==2021-10-27T12:00:00Z \
    Link:$CONTEXT_LINK
```

> Sample payload returned showing the temporal evolution of temperature and humidity properties:

Reviewer: add one example with `temporalValues`

    ],
    "type": "BeeHive"
}
```

* Let's update again the temperature property of the beehive:

```shell
http PATCH http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs \
    Link:$CONTEXT_LINK < beehive_secondTemperatureUpdate.json
```
beehive_secondTemperatureUpdate.json

```json
{
   "temperature": {
     "type": "Property",
     "value": 100,
     "unitCode": "CEL",
     "observedAt": "2020-05-10T10:20:30.98601Z",
     "observedBy": {
        "type": "Relationship",
        "object": "urn:ngsi-ld:Sensor:01"
     }
   }
}
```

* We can get the simplified temporal representation of the temperature property of the beehive:

```shell
http http://localhost:8080/ngsi-ld/v1/temporal/entities/urn:ngsi-ld:BeeHive:01?options=temporalValues \
    attrs==temperature \ 
    timerel==between \
    time==2019-10-25T12:00:00Z \
    endTime==2020-10-27T12:00:00Z \
    Link:$CONTEXT_LINK
```

The sample payload returned showing the simplified temporal evolution of temperature and humidity properties:

```json
{
    "@context": [
        "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld"
    ],

>     "temperature": {

Reviewer: not sure this is the payload you'll have

Author: updated

```shell
http http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01  options==keyValues Link:$CONTEXT_LINK
```
Sample payload returned showing a reduced version of the entity BeeHive:

```json
{
    "id": "urn:ngsi-ld:BeeHive:01",
    "type": "BeeHive",
    "humidity": 60,
    "temperature": 22.2,
    "belongs": "urn:ngsi-ld:Apiary:01",
    "managedBy": "urn:ngsi-ld:Beekeeper:01",
    "location": {
        "type": "Point",
        "coordinates": [
            24.30623,
            60.07966
        ]
    },
    "@context": [
        "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld"
    ]
}
```

### Append Attributes on Entities

* Let's add a name to the created beehive:

```shell
http POST http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs \
    Link:$CONTEXT_LINK <<<'
{
   "name":{
      "type":"Property",
      "value":"BeeHiveSophia"
   }
}
'
```

* The recently added name property can be deleted:

```shell
http DELETE http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs/name Link:$CONTEXT_LINK
```

### Partial Attributes Updates

> * We can update the temperature with a value grater then 40, using the following query:

Reviewer: ```suggestion
* We can update the temperature with a value greater than 40, using the following query:
```

# Quickstart

This quickstart guide shows a real use case scenario of interaction with the API in an Apiculture context.

## Prepare your environment

The provided examples make use of the HTTPie command line tool (installation instructions: https://httpie.org/docs#installation)

> All requests are grouped in a Postman collection that can be found [here](https://github.com/stellio-hub/stellio-docs/blob/master/API_Quick_Start.postman_collection.json).

Reviewer: I'm not sure this link is gonna work. Did you try it locally?

Author: I putted it in with Master (preparing the merge). I tested it locally (with the feature branch in the link) and it is taking me to the github page. It is like an external link.

Author: it is missing folder 'collection'

For more details about how to import a Postman collection see https://learning.postman.com/docs/getting-started/importing-and-exporting-data/.

Export the link to the JSON-LD context used in this use case in an environment variable for easier referencing in
the requests:

````shell
export CONTEXT_LINK="<https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld>; rel=http://www.w3.org/ns/json-ld#context; type=application/ld+json"
````

## Starting the Stellio Context Broker

> The Stellio Contex Broker source code can be found here : https://github.com/stellio-hub/stellio-context-broker

Reviewer: ```suggestion
The Stellio Context Broker source code can be found here : https://github.com/stellio-hub/stellio-context-broker
```

> To Start a Stellio instance. You can clone the stellio-context-broker and use the provided Docker compose configuration to run containers:

Reviewer: ```suggestion
To Start a Stellio instance. You can clone the Stellio repository and use the provided Docker compose configuration to run containers:
```

```shell
http GET https://raw.githubusercontent.com/stellio-hub/stellio-context-broker/develop/docker-compose.yml
```
and then start the container with:
```shell

> docker-compose -f docker-compose.yml up -d && docker-compose -f docker-compose.yml logs -f

Reviewer: without the `.env` file, I don't think it will work.

Author: Ok.
We need to clone the project instead of getting only the docker compose file?, Or to get the .env file with the docker compose?

Author: I have made a proposition, tell me if it is ok.

```

## Case study

This case study is written for anyone who wants to get familiar with the API, we use a real example to make it more concrete.

We will create the following entities:

- Beekeeper

- Apiary

> - Sensor_temperature.jsonld

Reviewer: ```suggestion
- Sensor temperature
```

> - Sensor_humidity.jsonld

Reviewer: ```suggestion
- Sensor humidity
```

> - Beehive.jsonld

Reviewer: ```suggestion
- Beehive
```

        },
        "managedBy": {
            "type": "Relationship",
            "object": "urn:ngsi-ld:Beekeeper:01"
        },
        "location": {
            "type": "GeoProperty",
            "value": {
                "type": "Point",
                "coordinates": [
                    24.30623,
                    60.07966
                ]
            }
        },
        "temperature": {
            "type": "Property",
            "value": 22.2,
            "unitCode": "CEL",
            "observedAt": "2019-10-26T21:32:52.98601Z",
            "observedBy": {
                "type": "Relationship",
                "object": "urn:ngsi-ld:Sensor:01"
            }
        },
        "humidity": {
            "type": "Property",
            "value": 60,
            "unitCode": "P1",
            "observedAt": "2019-10-26T21:32:52.98601Z",
            "observedBy": {
                "type": "Relationship",
                "object": "urn:ngsi-ld:Sensor:02"
            }
        },
        "@context": [
            "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld"
        ]
    }
]
'
```

> ### Discovering Entities

Reviewer: ```suggestion
### Discover Entities
```

> ### Get Types and Type of Entities

Reviewer: we typically call this "discovery", it should be part of the title

> * All available types of entities can be retrived:

Reviewer: ```suggestion
* All available types of entities can be retrieved:
```

```shell
http http://localhost:8080/ngsi-ld/v1/types
```

Sample payload returned showing the current available entity types:

```json
{
    "id": "urn:ngsi-ld:EntityTypeList:e2d9b009-d2d3-45af-8adc-048a433a80a2",
    "type": "EntityTypeList",
    "typeList": [
        "https://ontology.eglobalmark.com/apic#Beekeeper",
        "https://ontology.eglobalmark.com/apic#BeeHive",
        "https://ontology.eglobalmark.com/apic#Apiary",
        "https://ontology.eglobalmark.com/egm#Sensor"
    ]
}
```

* To get more details about a type of entity, a Beekeeper for example:

```shell

> http http://localhost:8080/ngsi-ld/v1/entities type==Beekeeper Link:$CONTEXT_LINK

Reviewer: This is not the correct URL to get details about an entity type

```

Sample payload returned showing more details about the entity Beekeeper:

```json
{
    "id": "https://ontology.eglobalmark.com/apic#Beekeeper",
    "type": "EntityTypeInfo",
    "typeName": "Beekeeper",
    "entityCount": 1,
    "attributeDetails": [
        {
            "id": "https://uri.etsi.org/ngsi-ld/name",
            "type": "Attribute",
            "attributeName": "name",
            "attributeTypes": [
                "Property"
            ]
        }
    ]
}
```

> ### Get Entities

Reviewer: ```suggestion
### Query Entities
```

            24.30623,
            60.07966
        ]
    },
    "@context": [
        "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld"
    ]
}
```

### Append Attributes on Entities

* Let's add a name to the created beehive:

```shell
http POST http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs \
    Link:$CONTEXT_LINK <<<'
{
   "name":{
      "type":"Property",
      "value":"BeeHiveSophia"
   }
}
'
```

* The recently added name property can be deleted:

```shell
http DELETE http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs/name Link:$CONTEXT_LINK
```

### Partial Attributes Updates

* We can update the temperature with a value grater then 40, using the following query:

```shell
http PATCH http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs \
    Link:$CONTEXT_LINK <<<'
{
   "temperature": {
     "type": "Property",
     "value": 42,
     "unitCode": "CEL",
     "observedAt": "2019-10-26T22:35:52.98601Z",
     "observedBy": {
        "type": "Relationship",
        "object": "urn:ngsi-ld:Sensor:01"
     }
   }
}
'
```

> * We can also update the beehive humidity:

Reviewer: ```suggestion
* We can also update the humidity of the beehive:
```

            "type": "Property",
            "value": 60.0
        }
    ],
    "temperature": [
        {
            "instanceId": "urn:ngsi-ld:Instance:33642ff7-9b66-42c4-bcdf-9ca640cba782",
            "observedAt": "2019-10-26T22:35:52.986010Z",
            "type": "Property",
            "value": 42.0
        },
        {
            "instanceId": "urn:ngsi-ld:Instance:73226f04-1c47-49a6-ab88-976cb7493bea",
            "observedAt": "2019-10-26T21:32:52.986010Z",
            "type": "Property",
            "value": 22.2
        }
    ],
    "@context": [
        "https://raw.githubusercontent.com/easy-global-market/ngsild-api-data-models/master/apic/jsonld-contexts/apic-compound.jsonld"
    ]
}
```

* Let's update again the temperature property of the beehive:

```shell
http PATCH http://localhost:8080/ngsi-ld/v1/entities/urn:ngsi-ld:BeeHive:01/attrs \
    Link:$CONTEXT_LINK <<< '
{
   "temperature": {
     "type": "Property",
     "value": 100,
     "unitCode": "CEL",
     "observedAt": "2020-05-10T10:20:30.98601Z",
     "observedBy": {
        "type": "Relationship",
        "object": "urn:ngsi-ld:Sensor:01"
     }
   }
}
'
```

> * We can get the simplified temporal representation of the temperature property of the beehive by adding the field `option` with the value `temporalValues` with the paramters of the query:

Reviewer: ```suggestion
* We can get the simplified temporal representation of the temperature property of the beehive by adding the request parameter `option` with the value `temporalValues` in the query:
```
